src/mcl.jl

```diff
@@ -0,0 +1,214 @@
+                clu_ixs[i] = 0
+                clu_sizes[clu_ix] = 0
+                unassigned_count += 1
+            end
+        end
+    else
+        unassigned_count = 0
+    end
+
+    # recode clusters numbers to be in 1:N range (or 0:N if there's collapsed cluster)
+    clu_id_map = zeros(Int, length(clu_sizes))
+    next_clu_ix = 0
+    @inbounds for i in eachindex(clu_ixs)
+        old_clu_ix = clu_ixs[i]
+        if old_clu_ix > 0
+            new_clu_ix = clu_id_map[old_clu_ix]
+            if new_clu_ix == 0
+                next_clu_ix += 1
+                clu_ixs[i] = clu_id_map[old_clu_ix] = next_clu_ix
+            else
+                clu_ixs[i] = new_clu_ix
+            end
+        end
+    end
+    old_clu_sizes = clu_sizes
+    clu_sizes = zeros(Int, next_clu_ix)
+    for (old_clu_ix, new_clu_ix) in enumerate(clu_id_map)
+        if new_clu_ix > 0
+            clu_sizes[new_clu_ix] = old_clu_sizes[old_clu_ix]
+        end
+    end
+
+    clu_ixs, clu_sizes, unassigned_count
+end
+
+# adjacency matrix expansion (matrix-wise raising to a given power) kernel
+# FIXME `_mcl_expand!()` that does not allocate new expanded matrix
+function _mcl_expand(src::Matrix, expansion::Integer)
+    src ^ expansion
+end
+
+# adjacency matrix expansion (matrix-wise raising to a given power) kernel
+# FIXME `_mcl_expand!()` that does not allocate new expanded matrix
+function _mcl_expand(src::Matrix, expansion::Number)
+    # we have to implement the workarond for matrix-power because of julia bug #16930
+    if isinteger(expansion)
+        return _mcl_expand(src, Integer(real(expansion)))
+    end
+    v, X = eig(src)
+    (eltype(v) <: Complex) || (any(v.<0) && (v = complex(v)))
+    Xinv = ishermitian(src) ? X' : inv(X)
+    scale(X, v.^expansion)*Xinv
+end
+
+# adjacency matrix inflation (element-wise raising to a given power) kernel
+function _mcl_inflate!(dest::Matrix{Float64}, src::Matrix{Complex128}, inflation::Number)
+    src_norm = vecnorm(src)
+    min_rel = -1E-3*src_norm
+    min_img = 1E-3*src_norm
+    @inbounds for (i, el) in enumerate(src)
+        rel = real(el)
+        img = imag(el)
+        if rel < min_rel || (abs(img) > min_img && abs(img) > 1E-3*abs(rel))
+            throw(InexactError())
+        end
+        dest[i] = max(0.0, rel)^inflation
+    end
+    return dest
+end
+
+# adjacency matrix inflation (element-wise raising to a given power) kernel
+function _mcl_inflate!(dest::Matrix{Float64}, src::Matrix{Float64}, inflation::Number)
+    src_norm = vecnorm(src)
+    min_el = -1E-3*src_norm
+    @inbounds for (i, el) in enumerate(src)
+        if el < min_el
+            throw(InexactError())
+        end
+        dest[i] = max(0.0, el)^inflation
+    end
+    return dest
+end
+
+""" mcl(adj::Matrix)
+
+    Identifies clusters in the weighted graph specified by its adjacency
+    matrix using Markov Clustering Algorithm (MCL).
+
+    Returns `MCLResult`.
+
+    See http://micans.org/mcl/
```

Comment:
Should unindent all lines except for the function signature. Also, the description of what it does should be in the imperative, e.g. "_Identify_ clusters in the weight graph" (note "identify" versus "identifies").

Rather than the line `Returns MCLResult`, you could make the function signature
```julia
    mcl(adj::Matrix{Float64}) -> MCLResult
```
Having the eltype in there as above would be helpful.